src/tpm2-tss-engine-common.c

```diff
@@ -588,3 +588,71 @@ init_tpm_key (ESYS_CONTEXT **esys_ctx, ESYS_TR *keyHandle, TPM2_DATA *tpm2Data)
     esys_ctx_free(esys_ctx);
     return r;
 }
+
+
+/** Initialize the ESYS TPM connection, load a key then evict the key
+ *
+ * Establish a connection with the TPM using ESYS libraries, create a primary
+ * key using the persistent handle, evict the key into persistent storage and
+ * return the result.
+ * @param tpm2Data The key data, owner auth and key auth to be used
+ * @param persKeyHandle The persistent handle of the key
+ * @param owner The handle of the keys owner
+ * @retval TSS2_RC_SUCCESS on success
+ * @retval TSS2_RCs according to the error
+ */
+int tpm2tss_load_and_evict(TPM2_DATA *tpm2Data, ESYS_TR persKeyHandle, ESYS_TR owner)
```

Comment:
If I understand correctly, this function does the same thing as `init_tpm_key`, except that it _also_ evicts key into  persistent storage:

`init_tpm_key` -> create key and return transient handle  
`tpm2tss_load_and_evict` -> create key and return persistent handle

Two questions:

1. (CC @zanebeckwith)  Do we need to evict into persistent storage?  Why not just [regenerate](https://github.com/tpm2-software/tpm2-tss/issues/430#issuecomment-309259556) the key each time?

1. `init_tpm_key` is not publicly exported, so it feels dubious to export it this way. Instead, perhaps there should be additional variants of `tpm2tss_{ecc,rsa}_{genkey,makekey}` that return a persistent handle?

1. If this exact function is needed, it should be renamed to be an obvious parallel of `init_tpm_key`.  Something like `init_tpm_key_persistant`.

Reply:
My thought with this is to mirror the concept of saving the key (an encrypted blob) to a file: yes, you could just go back through all the steps each time you want to use this key, but the more natural flow is to create it, save it in some way you can reuse it, then just use it without having to regenerate it. With saving it to a file, you have to first `load` it from the file, whereas if it's been "saved" to the TPM's nonvolatile memory you just need the handle where it's been saved.

This mirrors how the ecdaa and xtt libraries work, too. They assume the key has been created and require just the handle to be able to use the key.

Regenerating the key each time is mostly a pain for non-primary keys (you have to regenerate the primary key, and all intermediate keys in the tree, then finally the key you want). My understanding is that's the reason for the `evict`: to pin keys you're using so you don't have to deal with regenerating everything all the time

Reply:
But, you've raised an interesting point: for our usage, and considering this has to be something we have to submit upstream, maybe we just use a primary key for our authentication, and regenerate it every time we need it.

Honestly, I originally was thinking we would have to do quite a bit on top of what the engine provided, so adding the evict code would be only part of it. But perhaps we can just avoid that...

Reply:
Hmm, the more I think about it, I now remember at least one of our reasons for not going this route with ecdaa and xtt: if we plan to regenerate the key each time we need it, we open the possibility of the TPM running out of memory and thus not being able to regenerate the key.

By evicting the key into nonvolatile memory, we ensure it's always there, no matter what other uses the TPM might be put to.

Reply:
> we open the possibility of the TPM running out of memory and thus not being able to regenerate the key.

This is now solved by the in-kernel resource manager (/dev/tpmrmX), right?

Reply:
The in-kernel RM was released in kernel 4.11, I think: https://github.com/tpm2-software/tpm2-abrmd

Is that too-recent for some of our possible use cases?

Reply:
I don't think so.

The resource manager is only necessary if the TPM is used by multiple applications.  I can't envision a use case for us with an old kernel (pre-4.11) and multiple applications using the TPM.

- The router card is 5.4.
- The current popular third-party distro on IoT gateways is Ubuntu 18.04, which is 4.15.
- The IoT gateways with custom distros (some Advantech, MultiTech, SysTech, EuroTech) have a wide variety of kernel versions. But they don't currently have TPMs, so don't have other applications that would be using it.

Reply:
I've opened issue xaptum/enftun#124 to continue this dicussion.